code/messaging/src/phone/java/com/adobe/marketing/mobile/Messaging.java

```diff
@@ -191,20 +203,128 @@ public static void refreshInAppMessages() {
     public static void getPropositionsForSurfaces(@NonNull final List<Surface> surfaces, @NonNull final AdobeCallback<Map<Surface, List<Proposition>>> callback) {
+        if (callback == null ) {
+            Log.warning(LOG_TAG, CLASS_NAME, "Cannot get propositions as the provided callback is null.");
+            return;
+        }
 
+        if (surfaces == null || surfaces.isEmpty()) {
+            Log.warning(LOG_TAG, CLASS_NAME, "Cannot get propositions as the provided list of surfaces is null or empty.");
+            return;
+        }
+
+        final List<String> validSurfacePaths = new ArrayList<>();
+        for (final Surface surface : surfaces) {
+            if (surface.isValid()) {
+                validSurfacePaths.add(surface.toEventData());
+            }
+        }
+
+        if (validSurfacePaths.isEmpty()) {
+            Log.warning(LOG_TAG, CLASS_NAME, "Cannot get propositions as the provided list of surfaces has no valid items.");
+            return;
+        }
+
+        final Map<String, Object> eventData = new HashMap<>();
+        eventData.put(GET_PROPOSITIONS_EVENT, true);
+        eventData.put(SURFACES, validSurfacePaths);
+
+        final Event getPropositionsEvent = new Event.Builder(GET_PROPOSITIONS,
+                EventType.MESSAGING, EventSource.REQUEST_CONTENT)
+                .setEventData(eventData)
+                .build();
+
+        MobileCore.dispatchEventWithResponseCallback(getPropositionsEvent, 2000, new AdobeCallbackWithError<Event>() {
+            @Override
+            public void fail(final AdobeError adobeError) {
+                failWithError(callback, adobeError);
+            }
+
+            @Override
+            public void call(final Event event) {
+                try {
+                    final Map<String, Object> eventData = event.getEventData();
+                    if (MapUtils.isNullOrEmpty(eventData)) {
+                        failWithError(callback, AdobeError.UNEXPECTED_ERROR);
+                        return;
+                    }
+
+                    if (eventData.containsKey(RESPONSE_ERROR)) {
+                        final int errorCode = DataReader.getInt(eventData, RESPONSE_ERROR);
+                        failWithError(callback, convertToAdobeError(errorCode));
+                        return;
+                    }
+
+                    final List<Proposition> propositions = new ArrayList<>();
+                    final HashMap<Surface, List<Proposition>> requestedPropositions = new HashMap<>();
+
+                    final List<Map<String, Object>> retrievedPropositions = DataReader.optTypedListOfMap(Object.class, eventData, PROPOSITIONS, Collections.emptyList());
+                    if (retrievedPropositions == null || retrievedPropositions.isEmpty()) {
+                        failWithError(callback, AdobeError.UNEXPECTED_ERROR);
+                        return;
+                    }
+
+                    for (final Surface surface : surfaces) {
```

Comment:
This will not yield correct result. These are API provided surfaces, you might not have received responses for all of these from the server. You should read scope from the corresponding retrieved proposition to create the map.